In [18]:
# import
import torch
from torch.autograd import Variable
import numpy as np

In [19]:
# creating a variable
x = Variable(torch.ones(5))
print(x)
print(x.grad)

Variable containing:
 1
 1
 1
 1
 1
[torch.FloatTensor of size 5]

None


In [20]:
y = x+2
print(y, type(y))

Variable containing:
 3
 3
 3
 3
 3
[torch.FloatTensor of size 5]
 <class 'torch.autograd.variable.Variable'>


In [21]:
print(y.grad_fn)
print(y.grad)

None


In [25]:
z = y*y*3
print(z, type(z))

Variable containing:
 27
 27
 27
 27
 27
[torch.FloatTensor of size 5]
 <class 'torch.autograd.variable.Variable'>


In [26]:
out = z.mean()
print(out)

Variable containing:
 27
[torch.FloatTensor of size 1]



In [27]:
# doing backpropagation
out.backward()
# It will throw error - RuntimeError: there are no graph nodes that require computing gradients
# if not created with requires_grad=True
# so if want to avoid the warning, create x with requires_grad=True

RuntimeError: there are no graph nodes that require computing gradients

In [28]:
x = Variable(torch.ones(5), requires_grad=True)
print(x, type(x))

Variable containing:
 1
 1
 1
 1
 1
[torch.FloatTensor of size 5]
 <class 'torch.autograd.variable.Variable'>


In [30]:
y = x*2
z = y*y*y
out = z.mean()

print(x, y, z, out)

Variable containing:
 1
 1
 1
 1
 1
[torch.FloatTensor of size 5]
 Variable containing:
 2
 2
 2
 2
 2
[torch.FloatTensor of size 5]
 Variable containing:
 8
 8
 8
 8
 8
[torch.FloatTensor of size 5]
 Variable containing:
 8
[torch.FloatTensor of size 1]



In [31]:
# doing backpropagation
out.backward()

In [36]:
print(out.grad_fn,"\t\t", out.grad)
print(z.grad_fn,"\t\t", z.grad)
print(y.grad_fn,"\t\t", y.grad)

<torch.autograd.function.MeanBackward object at 0x0000000008909238> 		 None
<torch.autograd.function.MulBackward object at 0x0000000008909330> 		 None
<torch.autograd.function.MulConstantBackward object at 0x0000000007FE2DD8> 		 None


In [37]:
print(x.grad_fn,"\t\t", x.grad)

None 		 Variable containing:
 4.8000
 4.8000
 4.8000
 4.8000
 4.8000
[torch.FloatTensor of size 5]



$$ out = \frac{1}{5}(2x)^3 = \frac{8x^3}{5} $$  
$$ \frac{\partial(out)}{\partial x} = \frac{24x^2}{5} $$  

$$ \frac{\partial (out)}{\partial x_i}\bigr\rvert_{x_i=1} = \frac{24}{5} = 4.8 $$